circuits-lib/src/bridge_circuit/structs.rs

```diff
@@ -57,16 +57,108 @@ pub struct WatchtowerInput {
 }
 
 impl WatchtowerInput {
+    const FIRST_FIVE_OUTPUTS: usize = 5;
+    const NUMBER_OF_ASSERT_TXS: usize = 33;
+
     pub fn new(
         watchtower_idx: u8,
         watchtower_challenge_input_idx: u8,
-        watchtower_challenge_utxos: Vec<CircuitTxOut>,
-        watchtower_challenge_tx: CircuitTransaction,
-        watchtower_challenge_witness: CircuitWitness,
+        watchtower_challenge_utxos: Vec<TxOut>,
+        watchtower_challenge_tx: Transaction,
+        watchtower_challenge_witness: Witness,
+    ) -> Result<Self, &'static str> {
+        if watchtower_idx >= NUM_OF_WATCHTOWERS {
+            return Err("Watchtower index out of bounds");
+        }
+
+        let watchtower_challenge_tx = CircuitTransaction::from(watchtower_challenge_tx);
+
+        let watchtower_challenge_witness: CircuitWitness =
+            CircuitWitness::from(watchtower_challenge_witness);
+
+        let watchtower_challenge_utxos: Vec<CircuitTxOut> = watchtower_challenge_utxos
+            .into_iter()
+            .map(CircuitTxOut::from)
+            .collect::<Vec<CircuitTxOut>>();
+
+        Ok(Self {
+            watchtower_idx,
+            watchtower_challenge_input_idx,
+            watchtower_challenge_utxos,
+            watchtower_challenge_tx,
+            watchtower_challenge_witness,
+        })
+    }
+
+    pub fn from_txs(
+        kickoff_tx: &Transaction,
+        watchtower_tx: Transaction,
+        previous_txs: Option<&[Transaction]>,
     ) -> Result<Self, &'static str> {
+        let kickoff_txid = kickoff_tx.compute_txid();
+
+        let watchtower_challenge_input_idx = watchtower_tx
+            .input
+            .iter()
+            .position(|input| input.previous_output.txid == kickoff_txid)
+            .map(|ind| ind as u8)
+            .expect("Kickoff txid not found in watchtower inputs");
+
+        let output_index = watchtower_tx.input[watchtower_challenge_input_idx as usize]
+            .previous_output
+            .vout as usize;
+
+        let result = output_index
+            .checked_sub(Self::FIRST_FIVE_OUTPUTS + Self::NUMBER_OF_ASSERT_TXS)
+            .ok_or("Output index underflow")?
+            / 2;
+
+        let watchtower_idx = u8::try_from(result).map_err(|_| "Watchtower index too large")?;
```

Comment:
I think this try_from error is covered by the below check, we can do the check below then try_from + except